File: 백준/Platinum/14942. 개미/개미.py
# 개미
import sys
input = sys.stdin.readline
import math

def dfs(n):
    
    visited[n] = 1

    for i, cost in graph[n]:
        if visited[i]:
            continue
        else:
            dp[i][0] = n
            dist[i][0] = cost
            dfs(i)

n = int(input())
LOG = int(math.log(n, 2))
energy = []
visited = [0] * (n+1)
graph = [[] for _ in range(n+1)]
dp = [[0] * (LOG+1) for _ in range(n+1)] # dp[i][j]는 노드i의 2^j번째 부모노드
dist = [[0] * (LOG+1) for _ in range(n+1)] # 2^j 부모노드까지 거리
dist[1][0] = 1000001

for i in range(n):
    energy.append(int(input()))

for i in range(n-1):
    n1, n2, cost = map(int,input().split())
    graph[n1].append((n2, cost))
    graph[n2].append((n1, cost))

dfs(1)

for i in range(1, LOG+1):
    for j in range(1, n+1):
        dp[j][i] = dp[dp[j][i-1]][i-1]
        dist[j][i] = dist[j][i-1] + dist[dp[j][i-1]][i-1]

for i in range(1, n+1):
    cur = i
    cur_energy = energy[cur-1]
    
    for j in range(LOG, -1, -1):
        if dist[cur][j] <= cur_energy:
            cur_energy -= dist[cur][j]
            cur = dp[cur][j]

        if cur == 1:
            break

    print(cur)

File: 백준/Platinum/14942. 개미/개미.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int LOG;
    static ArrayList<int[]>[] graph;
    static int[] energy;
    static int[][] parent; // parent[i][j] -> i노드의 2^j 번째 부모노드
    static int[][] dist;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        LOG = (int) (Math.log(N) / Math.log(2));
        energy = new int[N+1];
        parent = new int[N+1][LOG+1];
        dist = new int[N+1][LOG+1];
        graph = new ArrayList[N+1];
        visited = new boolean[N+1];
        // 1번방이 시작이므로 부모노드는 없다. 따라서 거리는 임의의 큰 값으로 초기화
        dist[1][0] = 1000001;
        
        for (int i = 1; i <= N; i++) { // energy, graph 배열 초기화
            graph[i] =  new ArrayList<>();
            energy[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 0; i < N-1; i++) { // 거리 정보 입력, 저장
            st = new StringTokenizer(br.readLine());
            int n1 = Integer.parseInt(st.nextToken());
            int n2 = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            graph[n1].add(new int[] {n2, cost});
            graph[n2].add(new int[] {n1, cost});
        }

        dfs(1); // dfs로 노드별 바로 위의 부모, 거리 찾기

        for (int i = 1; i < LOG+1; i++) { // 희소행렬 계산
            for (int j = 1; j < N+1; j++) {
                int nParent = parent[j][i-1];
                parent[j][i] = parent[nParent][i-1];
                dist[j][i] = dist[j][i-1] + dist[nParent][i-1];
            }
        }

        for (int i = 1; i < N+1; i++) {
            int cur = i;
            int curEnergy = energy[cur];

            for (int j = LOG; j >= 0; j--) {
                if (curEnergy >= dist[cur][j]) {
                    curEnergy -= dist[cur][j];
                    cur = parent[cur][j];
                }

                if (cur == 1) { // 1번방에 도착하면 종료
                    break;
                }
            }
            sb.append(cur).append("\n");
        }

        System.out.println(sb);
    }

    static void dfs(int n) {
        visited[n] = true;

        for (int[] i : graph[n]) {
            int next = i[0];
            int cost = i[1];

            if (visited[next]) {
                continue;
            } else {
                parent[next][0] = n;
                dist[next][0] = cost;
                dfs(next);
            }
        }
    }
}
